initial implementation for stream readings (ws only)

edaniszewski opened the pull request:
This PR:
- adds a `ReadStream` function to the client interface
- adds support for reading a stream of reading data for the websocket client only (not yet supported in http api)
- creates a helper function for parsing response messages, since it is used by both the stream request and regular request

fixes #76


I don't think that this PR is quite ready yet, but opening for initial review. The issue that I'd like to solve prior to calling this ready is figuring out how to stop reading from a stream. Right now it just goes indefinitely until the program is exited. I think there are a number of ways this could be done...

#### exit callback

this would change the function signature
```go
ReadStream(scheme.ReadStreamOptions, chan<- *scheme.Read, exit func() bool) error
```

where the `exit` func would be called in the loop and when set to `true`, it would exit

```go
exiter = func() bool {
    time.Sleep(5 * time.Second)
    return true
}
```

#### exit channel - bool

this would change the function signature
```go
ReadStream(scheme.ReadStreamOptions, chan<- *scheme.Read, exit chan bool) error
```

where the `exit` channel would be read from in the loop and when passed `true`, it would exit

```go
exiter <- true
```

#### exit channel - closer

this would change the function signature
```go
ReadStream(scheme.ReadStreamOptions, chan<- *scheme.Read, exit chan struct{}) error
```

where the `exit` channel would be checked in the loop to see if it is open/closed. if closed, it would exit the loop.

```go
close(exiter)
```

-----------------

I feel like there probably are other approaches too, but the three above are the ones that came to me immediately. I think any option could work, so I guess its just a matter of semantics. The first option might be the least flexible, so I'd probably stick with one of the other two, but I'm interested in opinions/ideas

hoanhan101 commented:
I agree that having a way to stop a reading stream is necessary. From these 3 approaches above, I like the exit channel using bool one the most as it gets the work done and provides the cleanest api.

edaniszewski commented:
@hoanhan101 -- thanks for your feedback! I think that makes sense. I was kinda leaning towards that, but wasn't totally sure.

I think it definitely feels more intuitive to send `close <- true` or something like that. The bit that I'm less sure about from a user experience perspective is that if we use this approach (sending `true`), the user is still responsible for closing the channel.. which is fine, but is just another step for them to remember.

```go
readings = make(chan *scheme.Read)
exiter = make(chan bool)
go func() {
    if err := client.ReadStream(opts, readings, exiter); err != nil {
        log.Fatal(err)
    }
}()

// do something with the read stream. here we just wait for a bit and
// then close it for the example
time.Sleep(5 * time.Second)
exiter <- true

// this could/should also be deferred earlier..
close(exiter)
```

alternatively, if we just use the close semantic, I feel like it makes it harder to forget that the channel needs to be closed?

```go
readings = make(chan *scheme.Read)
exiter = make(chan bool)
defer close(exiter)

go func() {
    if err := client.ReadStream(opts, readings, exiter); err != nil {
        log.Fatal(err)
    }
}()

// do something with the read stream. here we just wait for a bit and
// then close it for the example
time.Sleep(5 * time.Second)
```


After writing all this, I think I might actually be leaning towards just using `close`.. I feel like if the channel is only ever used to send a boolean value, you can effectively get the same info just by closing the channel, which I think makes things simpler? 

if the usability doesn't feel right, we can always change it. I don't think it would be too difficult.

edaniszewski commented:
@hoanhan101 I updated to include the stop function and added a simple example program showcasing its usage. 

I still need to add tests around this, but I am going to try to fix https://github.com/vapor-ware/synse-sdk/issues/439 first, as I believe that needs to be fixed before any kind of integration test would pass.

edaniszewski commented:
I added tests.. but they aren't pretty. The integration test was failing due to a race condition. I think I am going to hold off on this PR until I can fix up the tests. Its not clear to me if the data race is a bug in the code or in the test.

edaniszewski commented:
@hoanhan101 I believe I found the bug and fixed it in my latest commit. It was due to the fact that the readings stream was being closed by the readstream function, when it should be closed by the caller of the readstream function.

I also found a bug related to closing the stream -- the stream gets closed server-side by sending a ws request with arguments `stop: true`, but it has no response. the previous implementation of `makeRequest` also read in the response, so I had to break that up into a request function and response function. 

I updated tests and things look to be passing on my machine. I still don't think it is the cleanest / best implementation and that there is lots of room for improvement, but for a first implementation, I think this is good enough and works.

If you could re-review my last commit, I think this should be ready to go!

hoanhan101 commented:
The new changes look good to me!

hoanhan101 commented:
Sorry I closed and reopened it by mistake!

edaniszewski commented:
no worries! thanks for all the review and discussion on this one! (:
